relayer/workers/parachaincommitmentrelayer/beefy-listener.go

```diff
@@ -75,81 +93,122 @@ func (li *BeefyListener) onDone(ctx context.Context) error {
+			blockNumber := gethheader.Number.Uint64()
+			var beefyLightClientEvents []*beefylightclient.ContractNewMMRRoot
 
-			signedCommitment := &store.SignedCommitment{}
-			err := types.DecodeFromHexString(msg.(string), signedCommitment)
+			contractEvents, err := li.queryBeefyLightClientEvents(ctx, blockNumber, &blockNumber)
 			if err != nil {
-				li.log.WithError(err).Error("Failed to decode BEEFY commitment messages")
+				li.log.WithError(err).Error("Failure fetching event logs")
+				return err
 			}
+			beefyLightClientEvents = append(beefyLightClientEvents, contractEvents...)
+
+			if len(beefyLightClientEvents) > 0 {
+				li.log.Info(fmt.Sprintf("Found %d BeefyLightClient ContractNewMMRRoot events on block %d", len(beefyLightClientEvents), blockNumber))
+			}
+			li.processBeefyLightClientEvents(ctx, beefyLightClientEvents)
+		}
+	}
+}
+
+// processLightClientEvents matches events to BEEFY commitment info by transaction hash
+func (li *BeefyListener) processBeefyLightClientEvents(ctx context.Context, events []*beefylightclient.ContractNewMMRRoot) error {
+	for _, event := range events {
+
+		contractAbi, err := abi.JSON(strings.NewReader(string(beefylightclient.ContractABI)))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		eventUnpacked, err := contractAbi.Unpack("FinalVerificationSuccessful", event.Raw.Data)
+		if err != nil {
+			return err
+		}
+
+		relayChainBlockNumber := (eventUnpacked[1].(*big.Int)).Int64()
+
+		li.log.WithFields(logrus.Fields{
+			"relayChainBlockNumber": relayChainBlockNumber,
+			"ethereumBlockNumber":   event.Raw.BlockNumber,
+			"ethereumTxHash":        event.Raw.TxHash.Hex(),
+		}).Info("Witnessed a new MMRRoot event")
 
-			blockNumber := signedCommitment.Commitment.BlockNumber
+		li.log.WithField("blockNumber", relayChainBlockNumber).Info("Getting hash for relay chain block")
+		blockHash, err := li.relaychainConn.GetAPI().RPC.Chain.GetBlockHash(uint64(relayChainBlockNumber))
+		if err != nil {
+			li.log.WithError(err).Error("Failed to get block hash")
+			return err
+		}
+		li.log.WithField("blockHash", blockHash.Hex()).Info("Got blockhash")
+
+		// TODO this just queries the latest MMR leaf in the latest MMR and our latest parahead from the relaychain.
+		// we should ideally be querying the latest and last few leaves in the latest MMR until we find
+		// the first parachain block that has not yet been fully processed on ethereum,
+		// and then package and relay all newer heads/commitments together with their corresponding leaf
+		mmrProof := li.relaychainConn.GetMMRLeafForBlock(uint64(relayChainBlockNumber-1), blockHash)
+		allParaHeads, ourParaHead := li.GetAllParaheads(blockHash, OUR_PARACHAIN_ID)
 
+		ourParaHeadProof := createParachainHeaderProof(allParaHeads, ourParaHead)
+
+		messagePackets, err := li.extractCommitments(ourParaHead, mmrProof, ourParaHeadProof)
+		if err != nil {
+			li.log.WithError(err).Error("Failed to extract commitment and messages")
```

Comment:
Need to return `err`?

Reply:
https://github.com/Snowfork/polkadot-ethereum/pull/442/commits/de6c07799d7397e6deb13c4e1614e92b2bc8c3a7